File: typescript/238.product-of-array-except-self.ts
/*
 * @lc app=leetcode id=238 lang=typescript
 *
 * [238] Product of Array Except Self
 */

// @lc code=start
function productExceptSelf(nums: number[]): number[] {
  const hash: {
    [key: string]: number[];
  } = {};
  nums.map((value, index) => {
    hash[index.toString()] = nums.filter((element, j) => j !== index);
  });

  let array: number[] = [];
  Object.keys(hash).map((key) => {
    array.push(
      hash[key].reduce((product, currentValue) => product * currentValue)
    );
  });

  return array;
}
// @lc code=end

File: 238.product-of-array-except-self.py
#
# @lc app=leetcode id=238 lang=python3
#
# [238] Product of Array Except Self
#

# @lc code=start
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        output_array = [1] * len(nums)

        for i in range(1, len(nums)):
            output_array[i] = output_array[i-1] * nums[i-1]

        postfix = 1
        for i in range(len(nums) - 1, -1, -1):
            output_array[i] *= postfix
            postfix *= nums[i]
        return output_array
# @lc code=end

